File: PE11/PE11.java
import java.io.*;
import java.util.*;


public class PE11 {
	public static final int SIZE = 20;
	public static final int LEN = 4;
	
	public static void getinput(int matrix[][]) throws FileNotFoundException {
		Scanner myfile = new Scanner(new File("PE11.txt"));
		int j = 0;
		while (myfile.hasNextLine()) {
			String linearr[] = myfile.nextLine().split(" ");
			for (int i = 0; i < SIZE; i++) {
				matrix[j][i] = Integer.parseInt(linearr[i]);
			}
			j++;
		}
		myfile.close();
	}
	
	// The following function returns product in direction, or -1 if undef.	
	public static int product(int matrix[][], int len, int i0, int j0, int di, int dj){ // smallest it would return is 1.
		int s = 1;
		int num = 0;
		while (num < len) {
			int newi = i0 + di * num;
			int newj = j0 + dj * num;
			if (newi > 0 && newi < SIZE && newj > 0 && newj < SIZE) {
				s *= matrix[i0 + di * num][j0 + dj * num];
			}
			num++;
		}
		return s;
	}
	
	public static void main(String args[]) throws FileNotFoundException {
		int maxim = 0;
		int matrix[][] = new int[SIZE][SIZE];
		printmatrix(matrix);
		getinput(matrix);
		printmatrix(matrix);
		for (int i = 0; i < SIZE; i++) {
			for (int j = 0; j < SIZE; j++) {
				for (int di = -1; di <= 1; di++) {
					for (int dj = -1; dj <= 1; dj++) {
						if (!(di == 0 && dj == 0)) {
							int testprod = product(matrix, LEN, i, j, di, dj);
							maxim = testprod > maxim ? testprod : maxim;
						}
					}
				}
			}
		}
		System.out.println(maxim);
	}
	
	public static void printmatrix(int matrix[][]) {
		for (int i = 0; i  < SIZE; i++) {
			for (int j = 0; j < SIZE; j++) {
				System.out.print(matrix[i][j]);
				System.out.print(" ");
			}
			System.out.println();
		}
		System.out.println();
	}
}

File: PE11/PE11.py
SIZE = 20
LEN = 4

def printmatrix(matrix):
    for i in range(SIZE):
        for j in range(SIZE):
            print(str(matrix[i][j])+" ", end="")
        print("")
    print("")

def getinput(matrix):
    myfile = open("PE11.txt", "r")
    for j in range(SIZE):
        linearr = myfile.readline().strip().split(" ")
        for i in range(0, SIZE):
            matrix[j][i] = int(linearr[i])
    myfile.close()
    
def product(matrix, lenn, i0, j0, di, dj):
    s = 1
    num = 0
    while (num < lenn):
        newi = i0 + di * num
        newj = j0 + dj * num
        if (newi > 0 and newi < SIZE and newj > 0 and newj < SIZE):
            s *= matrix[i0 + di * num][j0 + dj * num]
        num += 1
    return s

maxim = 0
matrix = []
for i in range(SIZE):
    row = []
    for j in range(SIZE):
        row.append(0)
    matrix.append(row)

getinput(matrix)

for i in range(0, SIZE):
    for j in range(0, SIZE):
        for di in range(-1, 2):
            for dj  in range(-1, 2):
                if not (di == 0 and dj == 0):
                    maxim = max(maxim, product(matrix, LEN, i, j, di, dj))
print(maxim)


